gobblin-service/src/main/java/org/apache/gobblin/service/monitoring/FsFlowGraphMonitor.java

```diff
@@ -0,0 +1,134 @@
+      .put(ConfigurationKeys.HOCON_FILE_EXTENSIONS, ConfigurationKeys.DEFAULT_CONF_EXTENSIONS).build());
+
+  public FsFlowGraphMonitor(Config config, Optional<? extends FSFlowTemplateCatalog> flowTemplateCatalog,
+      AtomicReference<FlowGraph> graph, Map<URI, TopologySpec> topologySpecMap, CountDownLatch initComplete)
+      throws IOException {
+    Config configWithFallbacks = config.getConfig(FS_FLOWGRAPH_MONITOR_PREFIX).withFallback(DEFAULT_FALLBACK);
+    this.pollingInterval =
+        TimeUnit.SECONDS.toMillis(configWithFallbacks.getLong(ConfigurationKeys.FLOWGRAPH_POLLING_INTERVAL));
+    this.flowGraphPath = new Path(configWithFallbacks.getString(ConfigurationKeys.FLOWGRAPH_REPO_DIR));
+    this.observer = new PathAlterationObserver(flowGraphPath);
+    this.flowGraphHelper = new BaseFlowGraphHelper(flowTemplateCatalog, topologySpecMap, flowGraphPath.toString(),
+        configWithFallbacks.getString(ConfigurationKeys.FLOWGRAPH_BASE_DIR), configWithFallbacks.getString(ConfigurationKeys.JAVA_PROPS_EXTENSIONS),
+        configWithFallbacks.getString(ConfigurationKeys.HOCON_FILE_EXTENSIONS));
+    this.listener = new FSPathAlterationFlowGraphListener(flowTemplateCatalog, graph, flowGraphPath.toString(), this.flowGraphHelper);
+
+   this.flowGraph = graph;
+   this.initComplete = initComplete;
+
+    if (pollingInterval == ConfigurationKeys.DISABLED_JOB_CONFIG_FILE_MONITOR_POLLING_INTERVAL) {
+      this.pathAlterationDetector = null;
+    } else {
+      this.pathAlterationDetector = new PathAlterationObserverScheduler(pollingInterval);
+      Optional<PathAlterationObserver> observerOptional = Optional.fromNullable(observer);
+      this.pathAlterationDetector.addPathAlterationObserver(this.listener, observerOptional,
+          this.flowGraphPath);
+    }
+  }
+
+  @Override
+  protected void startUp()
+      throws IOException {
+  }
+
+  @Override
+  public synchronized void setActive(boolean isActive) {
+    log.info("Setting the flow graph monitor to be " + isActive + " from " + this.isActive);
+    if (this.isActive == isActive) {
+      // No-op if already in correct state
+      return;
+    } else if (isActive) {
+      if (this.pathAlterationDetector != null) {
+        log.info("Starting the " + getClass().getSimpleName());
+        log.info("Polling flowgraph folder with interval {} ", this.pollingInterval);
+        try {
+          this.pathAlterationDetector.start();
+          // Manually instantiate flowgraph when the monitor becomes active
+          this.flowGraphHelper.populateFlowGraphAtomically(this.flowGraph);
+          // Reduce the countdown latch
+          this.initComplete.countDown();
+          log.info("Finished populating FSFlowgraph");
+        } catch (IOException e) {
+          log.error("Could not initialize pathAlterationDetector due to error: ", e);
+        }
+      } else {
+        log.warn("No path alteration detector found");
+      }
+    }
+    this.isActive = isActive;
```

Comment:
Should this be moved into the else-if statement instead? Since no-op if already in correct state

Reply:
It's representing a generic callable so I think a short name is fine here, usually we want more descriptive names for contextual variables.

Reply:
Need to consider the scenario where both isActive == false and this.isActive == true, if it gets moved to the else statement it only considers isActive == true and this.isActive = false

Reply:
And the noop does the early return